File: app/src/main/java/com/ruideraj/secretelephant/match/MatchAdapter.java
package com.ruideraj.secretelephant.match;

import android.graphics.Typeface;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.ruideraj.secretelephant.Constants;
import com.ruideraj.secretelephant.R;
import com.ruideraj.secretelephant.contacts.Contact;

import java.util.List;

public class MatchAdapter extends RecyclerView.Adapter {

    private List<Contact> mContacts;
    private int[] mMatches;
    private boolean[] mMatchesShown;  // Array containing which participants have their match shown.
    private int mMode;

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.match_list_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        ViewHolder vh = (ViewHolder) holder;
        Contact contact = mContacts.get(position);

        vh.participant.setText(contact.getName());

        boolean showMatch = mMatchesShown[position];

        if(showMatch) {
            if(mMode == Constants.MODE_SANTA) {
                Contact recipient = mContacts.get(mMatches[position]);
                vh.recipient.setText(recipient.getName());
            }
            else if(mMode == Constants.MODE_ELEPHANT) {
                int orderPosition = mMatches[position] + 1; // Add 1 to account for 0-based index.
                vh.recipient.setText("" + orderPosition);
            }

            vh.recipient.setTypeface(null, Typeface.NORMAL);
        }
        else {
            vh.recipient.setText("?");
            vh.recipient.setTypeface(null, Typeface.BOLD);
        }

    }

    @Override
    public int getItemCount() {
        return mContacts == null ? 0 : mContacts.size();
    }

    public void setData(List<Contact> contacts, int[] matches, int mode) {
        mContacts = contacts;
        mMatches = matches;
        mMatchesShown = new boolean[contacts.size()];
        mMode = mode;
        notifyDataSetChanged();
    }

    public void setNewMatches(int[] newMatches) {
        mMatches = newMatches;
        notifyItemRangeChanged(0, mContacts.size());
    }

    private class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView participant;
        TextView recipient;

        public ViewHolder(View view) {
            super(view);
            participant = view.findViewById(R.id.match_name);
            recipient = view.findViewById(R.id.match_recipient);
            recipient.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int position = getAdapterPosition();
            mMatchesShown[position] = !mMatchesShown[position];
            notifyItemChanged(position);
        }
    }
}

File: app/src/main/java/com/ruideraj/secretelephant/match/MatchAdapter.kt
package com.ruideraj.secretelephant.match

import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.ruideraj.secretelephant.Mode
import com.ruideraj.secretelephant.R

class MatchAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var exchange: MatchExchange? = null
    private var matchesShown: BooleanArray? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.match_list_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val ex = exchange!!
        val vh = holder as ViewHolder
        val contact = ex.contacts[position]

        vh.participant.text = contact.name

        val showMatch = matchesShown!![position]
        if (showMatch) {
            if (ex.mode == Mode.SANTA) {
                val recipient = ex.contacts[ex.matches[position]]
                vh.recipient.text = recipient.name
            } else {
                val orderPosition = ex.matches[position] + 1 // + 1 to account for zero-indexing
                vh.recipient.text = "$orderPosition"
            }

            vh.recipient.setTypeface(null, Typeface.NORMAL)
        } else {
            vh.recipient.text = "?"
            vh.recipient.setTypeface(null, Typeface.BOLD)
        }
    }

    override fun getItemCount() = exchange?.contacts?.size ?: 0

    fun setData(ex: MatchExchange) {
        exchange = ex
        matchesShown = BooleanArray(ex.contacts.size)
        notifyDataSetChanged()
    }

    fun setNewMatches(ex: MatchExchange) {
        exchange = ex
        notifyItemRangeChanged(0, ex.contacts.size)
    }

    private inner class ViewHolder(itemView: View,
                             val participant: TextView = itemView.findViewById(R.id.match_name),
                             val recipient: TextView = itemView.findViewById(R.id.match_recipient))
        : RecyclerView.ViewHolder(itemView) {

        init {
            recipient.setOnClickListener {
                matchesShown?.let {
                    val position = adapterPosition
                    it[position] = !it[position]
                    notifyItemChanged(position)
                }
            }
        }
    }
}

